### PR TITLE
HPCC-21951 Use Basedn name to get Basedn in ESP ws_access, etc

### PR DESCRIPTION
**Pull request by wangkx**
1. The ws_access service only returns LDAP Basedn for Permissions
request;
2. Basedn name (not the Basedn itself) is used to identify a
Basedn in other ws_access requests;
3. The ws_access does not echo Basedn rtype, rtitle, prefix, etc;
4. Depreciate the PermissionsResetInput, PermissionAddInput, and
ResourceAddInput methods which were used only for legacy ECLWatch;
5. Fix 1 bug in the PermissionsReset method. For example, if the
Userarray contains 'user1,user2', the existing code only resets
the permssions for the user1;
6. Clean some code.

Signed-off-by: wangkx <kevin.wang@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by wangkx** [comment]
@RussWhitehead please review.

File: esp/services/ws_access/ws_accessService.cpp [hunk 1]
```diff
@@ -1600,45 +1634,22 @@ bool Cws_accessEx::onPermissions(IEspContext &context, IEspBasednsRequest &req,
...
-
 bool Cws_accessEx::onResources(IEspContext &context, IEspResourcesRequest &req, IEspResourcesResponse &resp)
 {
     try
     {
-        checkUser(context, req.getRtype(), req.getRtitle(), SecAccess_Read);
+        Owned<IEspDnStruct> basednReq = createDnStruct();
+        getBasednReq(context, req.getBasednName(), req.getBasedn(), req.getRtype(), req.getRtitle(), basednReq);
+
+        checkUser(context, basednReq->getRtype(), basednReq->getRtitle(), SecAccess_Read);
```
Comment (RussWhitehead) [inline esp/services/ws_access/ws_accessService.cpp]: why do you need to create a basednReq struct?  Can't you just get the rType and Rtitle from req?

**Review by RussWhitehead — COMMENTED** [review]
@wangkx A couple minor questions

File: esp/services/ws_access/ws_accessService.cpp [hunk 2]
```diff
@@ -199,6 +199,41 @@ CLdapSecManager* Cws_accessEx::queryLDAPSecurityManager(IEspContext &context)
...
+                dn->setRtitle(cur.getRtitle());
+                return;
+            }
+        }
+        throw MakeStringException(ECLWATCH_INVALID_INPUT, "BaseDN %s not found", name);
+    }
+
+    //before version 1.14
+    if (isEmptyString(basedn) || isEmptyString(rType) || isEmptyString(rTitle))
+        throw MakeStringException(ECLWATCH_INVALID_INPUT, "BaseDN not specified");
```
[the remark this answers is not in the document]
Comment (wangkx) [reply esp/services/ws_access/ws_accessService.cpp]: Ok, I will change that.

File: esp/services/ws_access/ws_accessService.cpp (hunk 1 above)
Comment (wangkx) [reply esp/services/ws_access/ws_accessService.cpp]: I prefer to use the structure which has three members. Right now, the getBasednReq() already has 6 input parameters. If the structure is not used, the getBasednReq() will have 8 input parameters.

**Comment by wangkx** [comment]
@RussWhitehead please review again.

**Review by RussWhitehead — APPROVED** [review]
@wangkx Looks good

**Comment by wangkx** [comment]
@richardkchapman this PR is ready to be merged to master.

**Pull request merged**